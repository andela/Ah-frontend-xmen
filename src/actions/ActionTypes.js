const ActionTypes = {
  FETCH_ARTICLES_FAILED: 'FETCH_ARTICLES_FAILED',
  FETCH_ARTICLES_SUCCESS: 'FETCH_ARTICLES_SUCCESS',

  PASSWORD_RESET_SUCCESS: 'PASSWORD_RESET_SUCCESS',
  PASSWORD_RESET_FAILED: 'PASSWORD_RESET_FAILED',
  GET_COMMENTS_SUCCESS: 'GET_COMMENTS_SUCCESS',
  GET_COMMENTS_FAIL: 'GET_COMMENTS_FAIL',
  CREATE_COMMENT_SUCCESS: 'CREATE_COMMENT_SUCCESS',
  CREATE_COMMENT_FAIL: 'CREATE_COMMENT_FAIL',
  REGISTER_SUCCESS: 'REGISTER_SUCCESS',
  REGISTER_FAILURE: 'REGISTER_FAILURE',
  AUTH_SUCCESS: 'AUTH_SUCCESS',
  AUTH_FAIL: 'AUTH_FAIL',
  DELETE_COMMENT_SUCCESS: 'DELETE_COMMENT_SUCCESS',
  DELETE_COMMENT_FAIL: 'DELETE_COMMENT_FAIL',
  UPDATE_COMMENT_SUCCESS: 'UPDATE_COMMENT_SUCCESS',
  UPDATE_COMMENT_FAIL: 'UPDATE_COMMENT_FAIL',
  CREATE_ARTICLE_SUCCESS: 'ARTICLE_SUCCESS',
  CREATE_ARTICLE_FALIURE: 'ARTICLE_FALIURE',
  EDIT_ARTICLE_SUCCESS: 'EDIT_ARTICLE_SUCCESS',
  EDIT_ARTICLE_FALIURE: 'EDIT_ARTICLE_FALIURE',
  DELETE_ARTICLE_SUCCESS: 'DELETE_ARTICLE_SUCCESS',
  BOOKMARK_SUCCESS: 'BOOKMARK_SUCCESS',
  UNBOOKMARK_SUCCESS: 'UNBOOKMARK_SUCCESS',
  BOOKMARK_FAILURE: 'BOOKMARK_FAILURE',
  BOOKMARK_LIST_FETCHED: 'BOOKMARK_LIST_FETCHED',
  BOOKMARK_LIST_EMPTY: 'BOOKMARK_LIST_EMPTY',
};

export default ActionTypes;

export const FACEBOOK_SIGNIN_SUCCESS = 'FACEBOOK_SIGNIN_SUCCESS';

export const FACEBOOK_SIGNIN_FAIL = 'FACEBOOK_SIGNIN_FAIL';

export const GOOGLE_SIGNIN_SUCCESS = 'GOOGLE_SIGNIN_SUCCESS';

export const GOOGLE_SIGNIN_FAIL = 'GOOGLE_SIGNIN_FAIL';
export const GET_PROFILE_SUCCEEDS = 'GET_PROFILE_SUCCEEDS';
export const EDIT_PROFILE_SUCCEEDS = 'EDIT_PROFILE_SUCCEEDS';
export const GET_PROFILE_FAIL = 'GET_PROFILE_FAIL';
export const EDIT_PROFILE_FAILS = 'EDIT_PROFILE_FAILS';
export const GET_PROFILE_BEGIN = 'GET_PROFILE_BEGIN';
export const EDIT_PROFILE_BEGIN = 'EDIT_PROFILE_BEGIN';

export const ARTICLE_FETCH_NOT_FOUND = 'ARTICLE_FETCH_NOT_FOUND';
export const ARTICLE_FETCH_SUCCESSFUL = 'ARTICLE_FETCH_SUCCESSFUL';

export const ARTICLE_LIKE_SUCCESSFUL = 'ARTICLE_LIKE_SUCCESSFUL';
export const ARTICLE_LIKE_FAILURE = 'ARTICLE_LIKE_FAILURE';
export const ARTICLE_UNLIKE_SUCCESSFUL = 'ARTICLE_UNLIKE_SUCCESSFUL';
export const ARTICLE_UNLIKE_FAILURE = 'ARTICLE_UNLIKE_FAILURE';
export const ARTICLE_DISLIKE_SUCCESSFUL = 'ARTICLE_DISLIKE_SUCCESSFUL';
export const ARTICLE_DISLIKE_FAILURE = 'ARTICLE_DISLIKE_FAILURE';
export const ARTICLE_UNDISLIKE_SUCCESSFUL = 'ARTICLE_UNDISLIKE_SUCCESSFUL';
export const ARTICLE_UNDISLIKE_FAILURE = 'ARTICLE_UNDISLIKE_FAILURE';
export const ARTICLE_FETCH_LIKES_SUCCESSFUL = 'ARTICLE_FETCH_LIKES_SUCCESSFUL';
export const RATE_SUCCESSFUL = 'RATE_SUCCESSFUL';

export const REPORT_ARTICLE_SUCCEEDS = 'REPORT_ARTICLE_SUCCEEDS';
export const REPORT_ARTICLE_FAILS = 'REPORT_ARTICLE_FAILS';

export const getProfileSucceeds = profile => ({
  type: GET_PROFILE_SUCCEEDS,
  payload: { profile },
});

export const getProfileFails = error => ({
  type: GET_PROFILE_FAIL,
  payload: { error },
});

export const editProfileFails = error => ({
  type: EDIT_PROFILE_FAILS,
  payload: { error },
});

export const editProfilesSucceeds = payload => ({
  type: EDIT_PROFILE_SUCCEEDS,
  payload,
});

export const getProfileBegin = () => ({
  type: GET_PROFILE_BEGIN,
});

export const editProfileBegin = () => ({
  type: EDIT_PROFILE_BEGIN,
});

export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAIL = 'LOGIN_FAIL';

export const reportArticleSuccess = payload => ({
  type: REPORT_ARTICLE_SUCCEEDS,
  payload,
});

export const reportArticleFails = error => ({
  type: REPORT_ARTICLE_FAILS,
  payload: { error },
});
