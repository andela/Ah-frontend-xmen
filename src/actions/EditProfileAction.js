import * as actions from './ActionTypes';
import { handleErrors } from './ProfileActions';
import { BASE_URL } from '../constants';


export default function editProfile(profileData) {
  const username = localStorage.getItem('username');
  const token = localStorage.getItem('token');

  return (dispatch) => {
    dispatch(actions.editProfileBegin());
    return fetch(`${BASE_URL}/profiles/${username}/edit`, {
      method: 'PUT',
      mode: 'cors',
      body: JSON.stringify(profileData),
      headers: {
        'content-type': 'application/json',
        authorization: `Bearer ${token}`,
      },

    })
      .then(handleErrors)
      .then(res => res.json())
      .then((json) => {
        dispatch(actions.editProfilesSucceeds(json.profile));
        return json.profile;
      })
      .catch(error => dispatch(actions.editProfileFails(error)));
  };
}
