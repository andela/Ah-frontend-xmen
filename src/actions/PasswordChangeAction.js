
import ActionTypes from './ActionTypes';

const passwordChangeAction = (newPassword, token) => (dispatch) => {
  let statusCode;
  return fetch(`https://ah-backend-xmen-staging.herokuapp.com/api/users/password-reset/${token}/`,
    {
      method: 'PUT',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        user: {
          password: newPassword,
        },
      }),
    })
    .then((res) => {
      const { status } = res;
      statusCode = status;
      return res.json();
    })
    .then((data) => {
      if (statusCode < 400) {
        dispatch({
          type: ActionTypes.PASSWORD_RESET_SUCCESS,
          payload: data.user['password-reset'],
        });
      } else {
        dispatch({
          type: ActionTypes.PASSWORD_RESET_FAILED,
          payload: data.errors.error[0],
        });
      }
    });
};

export default passwordChangeAction;
