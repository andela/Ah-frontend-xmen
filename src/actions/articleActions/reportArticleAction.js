import * as actions from '../ActionTypes';
import { handleErrors } from '../ProfileActions';
import { BASE_URL } from '../../constants';

function reportArticleAction(payload) {
  const token = localStorage.getItem('token');
  const slug = localStorage.getItem('slug');
  return (dispatch) => {
    dispatch(actions.reportArticleSuccess());
    return fetch(`${BASE_URL}/articles/${slug}/report`, {
      method: 'POST',
      mode: 'cors',
      body: JSON.stringify(payload),
      headers: {
        'content-type': 'application/json',
        authorization: `Bearer ${token}`,
      },
    })
      .then(handleErrors)
      .then(res => res.json())
      .then((data) => {
        if (data) {
          dispatch(actions.reportArticleSuccess);
        }
      })
      .catch(error => dispatch(actions.reportArticleFails(error)));
  };
}
export default reportArticleAction;
