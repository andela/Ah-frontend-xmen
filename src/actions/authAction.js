
import { BASE_URL } from '../constants';
import ActionTypes from './ActionTypes';

const token = localStorage.getItem('token');

const getAuthUserDetails = () => dispatch => fetch(`${BASE_URL}/user`, {
  method: 'GET',
  headers: {
    Authorization: `Bearer ${token}`,
  },
}).then(
  res => res.json(),
).then((data) => {
  if (data.username || data.email) {
    return fetch(`${BASE_URL}/profiles/${data.username}`, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
      },

    }).then(
      res => res.json(),
    ).then(
      (data) => {
        if (data.profile) {
          dispatch({
            type: ActionTypes.AUTH_SUCCESS,
            payload: data,
          });
        } else {
          dispatch({
            type: ActionTypes.AUTH_FAIL,
            payload: data.errors,
          });
        }
      },
    );
  }
});
export default getAuthUserDetails;
