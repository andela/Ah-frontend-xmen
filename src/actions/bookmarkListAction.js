import ActionTypes from './ActionTypes';

const { BASE_URL } = process.env;
const token = localStorage.getItem('token');

const bookmarkListing = slug => dispatch => fetch(`${BASE_URL}/bookmarks`, {
  headers: {
    'Content-Type': 'application/json',
    authorization: `Bearer ${token}`,
  },
}).then(response => response.json())
  .then((data) => {
    if (data.bookmarks) {
      dispatch({
        type: ActionTypes.BOOKMARK_LIST_FETCHED,
        payload: data.bookmarks,
        slug,
      });
    } else {
      dispatch({
        type: ActionTypes.BOOKMARK_LIST_EMPTY,
        message: 'You do not have any articles bookmarked',
      });
    }
  });

export default bookmarkListing;
