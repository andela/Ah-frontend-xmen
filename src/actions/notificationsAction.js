import { FETCH_NOTIFICATIONS_SUCCESS } from './ActionTypes';
import { BASE_URL } from '../constants';


const notificationsAction = markAsRead => (dispatch) => {
  let statusCode;
  const token = localStorage.getItem('token');
  return fetch(
    `${BASE_URL}/notifications/all`,
    {
      method: markAsRead ? 'PUT' : 'GET',
      headers: { Authorization: `Bearer ${token}` },
    },
  ).then((res) => {
    statusCode = res.status;
    return res.json();
  })
    .then((data) => {
      if (statusCode === 403) {
        localStorage.removeItem('token');
      } else if (statusCode < 400) {
        dispatch({
          type: FETCH_NOTIFICATIONS_SUCCESS,
          payload: data.notification,
        });
      }
    });
};

export default notificationsAction;
