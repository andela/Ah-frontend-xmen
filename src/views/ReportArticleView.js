/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReportArticle from '../components/Articles/ReportArticle';
import reportArticleAction from '../actions/articleActions/reportArticleAction';


export class ReportArticleView extends Component {
  constructor(props) {
    super(props);
    this.state = {
      reason: '',
    };
    this.handleInput = this.handleInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleInput(e) {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  handleSubmit(e) {
    const { history } = this.props;
    e.preventDefault();
    this.props.reportArticleAction(this.state);
    history.push('/article');
  }

  render() {
    return (
      <div>
        <ReportArticle
          onSubmit={this.handleSubmit}
          onChange={this.handleInput}
          report={this.state}
        />
      </div>
    );
  }
}

ReportArticleView.propTypes = {
  reportArticleAction: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
};


// const mapStateToProps = state => ({
//   reason: state.reportReducer.reason,
//   error: state.reportReducer.error,
// }
// );


export default ReportArticleView;
