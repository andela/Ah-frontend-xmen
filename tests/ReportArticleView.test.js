import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import expect from 'expect';
import * as ActionTypes from '../src/actions/ActionTypes';
import reportArticleAction from '../src/actions/articleActions/reportArticleAction';
import { ReportArticleView } from '../src/views/ReportArticleView';


const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('report article action', () => {
  let store;
  jest.useFakeTimers();
  beforeEach(() => {
    store = mockStore({});
    localStorage.clear();
  });
  it('should have a type of REPORT_ARTICLE_SUCEEDS', () => {
    const payload = 'the reason i report is';
    const expectedAction = {
      type: ActionTypes.REPORT_ARTICLE_SUCCEEDS,
      payload: 'the reason i report is',
    };
    const result = ActionTypes.reportArticleSuccess(payload);
    expect(result).toEqual(expectedAction);
  });
  it('should be able to report an article', () => {
    const expectedAction = (reportArticleAction);
    expect(reportArticleAction).toEqual(expectedAction);
  });
  it('should handle submit', () => {
    localStorage.setItem('slug', '123');
    const props = {
      reportArticleAction: jest.fn(),
      handleSubmit: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
    };
    const wrapper = shallow(<ReportArticleView {...props} />);
    wrapper.instance().handleSubmit(event);
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle submit with state != ''", () => {
    localStorage.setItem('slug', '123');
    const props = {
      reportArticleAction: jest.fn(),
      handleSubmit: jest.fn(),
    };
    const event = { preventDefault: jest.fn() };
    const wrapper = shallow(<ReportArticleView {...props} />);
    wrapper.setState({ reason: 'something' });
    wrapper.instance().handleSubmit(event);
    const spy = jest.spyOn(wrapper.instance().props, 'reportArticleAction');
    expect(spy).toHaveBeenCalled();
  });

  it('should call handleInput', () => {
    const props = {
    };
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'reason',
        value: 'contains violent content',
      },
    };

    const wrapper = shallow(<ReportArticleView {...props} />);
    wrapper.instance().setState = jest.fn();
    wrapper.instance().handleInput(event);
    expect(wrapper.instance().setState).toHaveBeenCalled();
  });

  it('should call handleCancel', () => {
    const props = {
    };
    localStorage.setItem('slug', '123');
    window.location.assign = jest.fn();
    const wrapper = shallow(<ReportArticleView {...props} />);
    wrapper.instance().handleCancel();
    expect(window.location.assign).toHaveBeenCalled();
  });
});
