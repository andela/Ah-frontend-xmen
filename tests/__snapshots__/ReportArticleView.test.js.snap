// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report article action should handle submit 1`] = `ShallowWrapper {}`;
